[Fleet] Enable package filtering in serverless using an exclude list

nchaulet opened the pull request:
## Summary

Resolve #166742 

Enable package filtering


## Todo

Test on security:
* [x] Verify endpoints, cspm packages are here

Test on observability:
* [x] Verify system, apm packages are here

elasticmachine commented:
Pinging @elastic/fleet (Team:Fleet)

apmmachine commented:
### :robot: GitHub comments

<details><summary>Expand to view the GitHub comments</summary>
<p>

Just comment with:
- `/oblt-deploy` : Deploy a Kibana instance using the Observability test environments.
- `/oblt-deploy-serverless` : Deploy a serverless Kibana instance using the Observability test environments.
- `run` `elasticsearch-ci/docs` : Re-trigger the docs validation. (use unformatted text in the comment!)

</p>
</details>

jlind23 reviewed (changes requested):
Holding off until APM, Endpoint and other packages are migrated

jlind23 commented:
@nchaulet @jsoriano After discussing this change with @andresrc I believe we could come up with a short term solution that will release the pressure on the integrations developer.
The solution will consist in two things:
- If packages are using spec V3, they should be available
- If packages are using spec V2 and are not part of the exclusion list, they should be available

We should have two exclusions lists (this as meant to be removed fairly soon), one for observability with all integration that should NOT be available and one for security serving the same purpose.

What do you think? These lists can be for now, hardcoded in Kibana.

cc @epixa for security related integrations

andresrc commented:
Thanks @jlind23 for proposing this option, it would be great to see if it's feasible. The only change would be:

> If packages are using spec V2 and are not part of the exclusion list, they should be available

If packages are using any spec version < V3...

As we have 63 packages in v1.

Thanks

jlind23 commented:
@mrodm @nchaulet @jsoriano @joshdover can we have your thoughts on this?

joshdover commented:
I replied to the email thread

nchaulet commented:
> @mrodm @nchaulet @jsoriano @joshdover can we have your thoughts on this?

In kibana it will be not possible to filter on spec version (we do not have that info until we load the package manifest) so if we go the exclusion list way, we probably want to exclude all the possible packages, and we will not be able to release a new package only for a solution without either updating that list or finishing the work to use spec filtering.

jsoriano commented:
> If packages are using spec V2 and are not part of the exclusion list, they should be available
>
> We should have two exclusions lists (this as meant to be removed fairly soon), one for observability with all integration that should NOT be available and one for security serving the same purpose.
>
> What do you think? These lists can be for now, hardcoded in Kibana.

With inclusion/exclusion lists, package developers should still do at least a minimal testing of their packages on serverles, what is in many cases the biggest part of the required effort. And as Nicolas mentioned, including or excluding packages would require PRs in Kibana.

Also, we would need to design and implement this new mechanism in short time, and keep it forever. Removing it after the release would be a breaking change, as some packages might disappear.

I think we should continue with the current plan. @jlind23 @andresrc I see how the short timing is unfortunate, but take into account that there are no hard deadlines to complete _all_ packages. I don't think it is a big issue if some packages are not ready till end of October, or even till end of year. Developers could focus now on testing and migrating the most used packages, and others could be migrated later, or when missed and requested by users. We are trading off the early availability of some packages for the certainty that they properly work in the new offerings.

We are also introducing filtering for some validation checks during this month, what will allow to delay for example the migration of some dashboards. And we are planning to introduce additional automation for formatting rules. Both things should alleviate migration of packages.

> If packages are using any spec version < V3...
> 
> As we have 63 packages in v1.

Please also notice that some of the checks introduced in v2 and v3 guide to the fix of actual bugs, so updating packages that are still using spec v1 brings more benefits apart of the opt-in for serverless.

jlind23 commented:
After discussing with @jsoriano this is where we are at:
- [ ] Integrations filtering based on package spec should be turned on for Security project as the migration work as been done. @epixa please keep me honest here.
- [ ] Integrations filtering should NOT be enable for Observability project but we should still filter out some packages and the list will be provided by @andresrc 
- [ ] The specific Observability filtering MUST be removed before Serverless going GA

Known limitation: Any update on the list will require a new Kibana release.

cc @nimarezainia @joshdover

nchaulet commented:
@elasticmachine merge upstream

joshdover commented:
> Integrations filtering based on package spec should be turned on for Security project as the migration work as been done.

We may actually not want to do this for Security projects either. Many o11y integrations are applicable to security and they will be filtered out. There are also some ML model packages that are specific to security that are not yet updated. @peteharverson

peteharverson commented:
> There are also some ML model packages that are specific to security that are not yet updated.

Security integrations containing ML components (models; jobs) which still need updating are:
- [ ] [ded](https://github.com/elastic/integrations/tree/main/packages/ded)
- [ ] [dga](https://github.com/elastic/integrations/tree/main/packages/dga)
- [ ] [lmd](https://github.com/elastic/integrations/tree/main/packages/lmd) 
- [ ] [problemchild](https://github.com/elastic/integrations/tree/main/packages/problemchild)

@ajosh0504 is currently looking at updating these. Did I miss any off the list?

peteharverson commented:
@andresrc note that the apache and nginx packages also contain ML anomaly detection jobs, and looks like they still need to be updated.

jlind23 commented:
@epixa @andresrc could you please provide the exclusion list for both project types? Would be great to have it in order to unblock @nchaulet.

ebeahan commented:
@jlind23 SEI-owned security integrations have all been migrated to spec v3 (cc @narph @andrewkroh), but we can't speak for security integrations owned by other groups in security: Endpoint, ML, CSPM.

I've shared this with the security team leads for visibility and asking others to voice if they need exclusions.

andresrc commented:
From the Observability side, we have no exclusions. The packages requiring special care are APM and synthetics, both of them are bundled and they will be updated to spec v3 /cc @ruflin @tommyers-elastic

nchaulet commented:
I updated the PR to go with spec filtering and capabilities for security and an exclusion list for observability that remove the security only and removed in 8.11 integrations

nchaulet commented:
Looks like `cloud_security_posture` is not in 3.0 yet

nchaulet commented:
@andresrc @peteharverson I updated the PR to use an exclude list for security and observability while packages are migrated to spec v3, if it's good for you I will move forward with that and merge that PR.

For now in oblt we will exclude
```
xpack.fleet.internal.registry.excludePackages: [
    # Security integrations
    'endpoint',
    'beaconing',
    'osquery_manager',
    # Removed in 8.11 integrations
    'cisco',
    'microsoft',
    'symantec',
    'cyberark',
  ]
  ```
  
  In security:
  ```
  xpack.fleet.internal.registry.excludePackages: [
    # Oblt integrations
    'apm',
    'synthetics',
    # Removed in 8.11 integrations
    'cisco',
    'microsoft',
    'symantec',
    'cyberark',
  ]
  ```

nchaulet commented:
@elasticmachine merge upstream

ebeahan commented:
> @andresrc @peteharverson I updated the PR to use an exclude list for security and observability while packages are migrated to spec v3, if it's good for you I will move forward with that and merge that PR.

@nchaulet @jlind23 are we _not_ limiting integrations to package-spec v3 in serverless now?

jlind23 commented:
@ebeahan unfortunately for now we are not as at least some observability integrations are still not migrated and it offers us a bit more time to make it happened.

nchaulet commented:
FYI I will merge that PR later today

terrancedejesus commented:
Not sure if this is helpful.
Regarding the prebuilt detection rules for the SIEM, we have not reviewed package registry v3 yet and are completely reliant on v2 at the moment. This is the [security_detection_engine](https://epr.elastic.co/search?package=security_detection_engine&prerelease=true&all=true&include_policy_templates=true) package. Please disregard my lack of knowledge at the moment, but is it required for us to update to v3 anytime soon?

nchaulet commented:
> Not sure if this is helpful.
Regarding the prebuilt detection rules for the SIEM, we have not reviewed package registry v3 yet and are completely reliant on v2 at the moment. This is the [security_detection_engine](https://epr.elastic.co/search?package=security_detection_engine&prerelease=true&all=true&include_policy_templates=true) package. Please disregard my lack of knowledge at the moment, but is it required for us to update to v3 anytime soon?

Updating to v3 will be required in the future to enable filtering based on packages capabilities the fleet team will communicate on the deadline for this soon.

To resume before merging we are starting with an exclusion list for both oferring and no filtering on package spec or capabilities (capabilities filtering require to filter on spec v3) until all package migrate to v3 and we will communicate a deadline date for migrating to v3.

pantea-elastic commented:
> > There are also some ML model packages that are specific to security that are not yet updated.
> 
> Security integrations containing ML components (models; jobs) which still need updating are:
> 
> * [x]  [ded](https://github.com/elastic/integrations/tree/main/packages/ded) [[Data Exfiltration Detection] Update files to add Serverless support to the package integrations#8108](https://github.com/elastic/integrations/pull/8108)
> * [ ]  [dga](https://github.com/elastic/integrations/tree/main/packages/dga)
> * [x]  [lmd](https://github.com/elastic/integrations/tree/main/packages/lmd)  [[Lateral Movement Detection] Update files to add Serverless support to the package integrations#8121](https://github.com/elastic/integrations/pull/8121)
> * [x]  [problemchild](https://github.com/elastic/integrations/tree/main/packages/problemchild) [[ProblemChild] Update files to add Serverless support to the package integrations#8106](https://github.com/elastic/integrations/pull/8106)
> 
> @ajosh0504 is currently looking at updating these. Did I miss any off the list?

After discussions with the ML and fleet teams, the DGA package has installation issues due to its size, and for now will not be opted in. ML team has opened [this issue](https://github.com/elastic/ml-team/issues/1059) ([slack thread](https://elastic.slack.com/archives/CNMNXV4RG/p1699443760669999?thread_ts=1698177871.711959&cid=CNMNXV4RG)) to resolve scaling to allow for bigger models.

susan-shu-c commented:
Hi @nchaulet had a quick question:

I'm in the Security org and the DGA package is on the exclude list per our discussions at the time.
The DGA package is on spec 2.2.0 and we are looking to update it to spec 3.0 per [the 9.0 issue](https://github.com/elastic/ingest-dev/issues/4484). 

I want to check if the spec version is being used for the exclude list logic? As in, for non-Serverless, if we update DGA to spec 3.0, it still is properly excluding on Serverless? Asking just in case.

nchaulet commented:
Hi @susan-shu-c, the spec version is not used for the exclude list logic, event after upgrading DGA to 3.0 it will still be excluded in serverles
